daemon/cmd/endpoint.go

```diff
@@ -453,6 +455,13 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
 		pod, k8sMetadata, err := d.fetchK8sMetadataForEndpoint(ep.K8sNamespace, ep.K8sPodName)
 		if err != nil {
 			ep.Logger("api").WithError(err).Warning("Unable to fetch kubernetes labels")
+		} else if types.UID(ep.K8sUID) != pod.GetUID() {
+			log.WithFields(logrus.Fields{
+				logfields.K8sPodName: ep.K8sNamespace + "/" + ep.K8sPodName,
+				logfields.K8sUID:     ep.K8sUID,
+			}).Warn("Detected outdated Pod UID during Endpoint creation. " +
+				"The Endpoint may be created with outdated identity due to delay in Kubernetes event updates, i.e. until Pod event is processed by Cilium. " +
+				"If the Endpoint persists with an outdated identity for an extended period of time, please report an issue.")
```

Comment:
I'm not sure if we should Warn in here, this is a transient error and we will recover afterwards.

Reply:
Wouldn't it be useful to know that the bugs we've hit with statefulsets were detected and Cilium is warning the user that there might be a period of time where the identity is out of date? It seems useful to me.

Reply:
+1 from me for warning, I think it's useful. 

Another way to prevent that would be to fail CNI ADD when we don't have a pod in the informer store and let Kubelet retry later on, but this has lots of other consequences. Or wait for a fixed amount of time for store to get updated pod and if it doesn't happen then fail CNI ADD. But this is probably quite a lot of work and outside of this PR's scope.

Reply:
Another place similar to this is metadataResolver https://github.com/cilium/cilium/blob/1518818397ac98be01573bb0c52cd0ece122f3f7/pkg/endpoint/endpoint.go#L1776
where we could get a "random" pod instead of the correct one in case the store is lagging.

How about an alternative approach:
- we try to get the pod from the store
- if it's not there or has incorrect UID we fetch it from k8s control plane directly like in high-scale case. Not ideal, but in most cases, it should not be triggered and it would ensure we have the correct pod information always.

Reply:
While you are correct, I don't see any negative consequences for having an out-of-date pod object reference in the `Endpoint`. In fact, we don't even update the reference on pod update events, so they are pretty much always stale after `metadataResolver` runs. This is because we don't use the metadata for anything else besides initially fetching the labels for identity purposes. Label changes are detected in the pod update events, but the reference to the object is never updated.

Reply:
> I don't see any negative consequences for having an out-of-date pod object reference

If we have reference to previous pod with incorrect UID then in endpoint synchronizer we will set ownerReference to incorrect pod that no longer exists:
https://github.com/cilium/cilium/blob/02d04b34005c45d955a6850d3faf561d9eae8f78/pkg/endpointmanager/endpointsynchronizer.go#L115
which means, kube-controller-manager will GC that CEP immediately.

Reply:
I've changed the code to fetch the latest pod object in the case that the UID is mismatching during endpoint creation.

Reply:
How will the user know that a pod has been successfully updated? Should we log a message indicating that a pod was stale for XXX minutes?

Also, error messages "aren't real", only alerts are. There should be a metric for whether or not an endpoint has a correct pod.

Reply:
There will be log msgs indicating what Cilium is doing when the outdated UID case is hit. We suspect this case to be very rare anyway.